File: cpp-examples/mutability.cpp
#include <vector>
#include <iostream>

void print_shopping_list(std::vector<std::string> const &shopping_list) {
    //                                            ^^^^^ explicitly immmutable
    if (shopping_list.size() == 0) {
        std::cout << "\t-- nothing\n";
    }

    for (const auto& item: shopping_list) {
        std::cout << "\t* " << item << '\n';
    }
}

void buy_groceries(std::vector<std::string> shopping_list) {
    //             ^^^^^^^^^^^^^^^^^^^^^^^^ mutable by default
    std::cout << "Going out to buy:\n";
    print_shopping_list(shopping_list);

    shopping_list.clear();

    std::cout << "things left to buy:\n";
    print_shopping_list(shopping_list);
}

int main() {
    std::vector<std::string> shopping_list {"Pasta", "Milk", "Toilet Paper"};
    shopping_list.push_back("Chocolate");
    //^^^^^^^^^^^^^^^^^^^^^ mutable by default

    buy_groceries(shopping_list);
}

File: rust-examples/src/mutability.rs
fn buy_groceries(shopping_list: Vec<&str>) {
    println!("Going out to buy: {:?}", shopping_list);

    shopping_list.clear();

    println!("Things left to buy: {:?}", shopping_list);
}

fn main() {
    let shopping_list: Vec<&str> = vec!["Pasta", "Milk", "Toilet Paper"];
    shopping_list.push("Chocolate");

    buy_groceries(shopping_list);
}